blood/src/main/java/cz/cuni/mff/d3s/blood/node_type_tracker/NodeTypeTrackerCollector.java

```diff
@@ -0,0 +1,78 @@
+package cz.cuni.mff.d3s.blood.node_type_tracker;
+
+import cz.cuni.mff.d3s.blood.report.Report;
+import cz.cuni.mff.d3s.blood.report.dump.ManualDump;
+import org.graalvm.compiler.graph.Node;
+import org.graalvm.compiler.nodes.StructuredGraph;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class NodeTypeTrackerCollector {
+    // the default of 16 doesn't fit even the most trivial programs
+    private static final int HASHMAP_INIT_CAPACITY = 64;
+
+    private static NodeTypeTrackerCollector instance = null;
+
+    private final ConcurrentHashMap<Class, ConcurrentHashMap<Class, NodeTrackerValue>> postPhaseTypes = new ConcurrentHashMap<>(HASHMAP_INIT_CAPACITY);
+    private final ConcurrentHashMap<Class, ConcurrentHashMap<Class, NodeTrackerValue>> prePhaseTypes = new ConcurrentHashMap<>(HASHMAP_INIT_CAPACITY);
+
+    public NodeTypeTrackerCollector() {
+        // FIXME implement dumping
+        Report.getInstance().registerDump(new ManualDump("nodetypematrix", () -> new byte[0]));
+    }
+
+    public static NodeTypeTrackerCollector getInstance() {
+        if (instance == null) {
+            instance = new NodeTypeTrackerCollector();
+        }
+        return instance;
+    }
+
+    public final void updateNodeTypes(StructuredGraph graph, ConcurrentHashMap<Class, NodeTrackerValue> phaseTable) {
+        HashMap<Class, Long> nodeCount = new HashMap<>();
+        for (Node n : graph.getNodes()) {
+            Long count = nodeCount.getOrDefault(n.getClass(), 0l);
```

Comment:
Typo

Reply:
I don't see it. If you mean the `0l` constant, that's type specification placed there on purpose...

Reply:
Sorry. In the font GitHub UI uses, it seemed to me like "zero one". I thought you also pressed 1 while aiming at 0.